crates/bevy_utils/src/intern.rs

```diff
@@ -0,0 +1,310 @@
+    /// Creates a static reference to a value equal to `self.borrow()`, possibly leaking memory.
+    fn leak(self) -> &'static T;
+}
+
+impl<T: ?Sized> Leak<T> for &'static T {
+    fn leak(self) -> &'static T {
+        self
+    }
+}
+
+impl<T: ?Sized> Leak<T> for Box<T> {
+    fn leak(self) -> &'static T {
+        Box::leak(self)
+    }
+}
+
+impl<T: Clone + ?Sized> Leak<T> for Cow<'static, T> {
+    fn leak(self) -> &'static T {
+        match self {
+            Cow::Borrowed(value) => value,
+            Cow::Owned(value) => Box::leak(Box::new(value)),
+        }
+    }
+}
+
+/// A thread-safe interner which can be used to create [`Interned<T>`] from `&T`
+///
+/// For details on interning, see [the module level docs](self).
+///
+/// The implementation ensures that two equal values return two equal [`Interned<T>`] values.
+pub struct Interner<T: ?Sized + 'static>(OnceLock<RwLock<HashSet<&'static T>>>);
+
+impl<T: ?Sized> Interner<T> {
+    /// Creates a new empty interner
+    pub const fn new() -> Self {
+        Self(OnceLock::new())
+    }
+}
+
+impl<T: Hash + Eq + ?Sized> Interner<T> {
+    /// Return the [`Interned<T>`] corresponding to `value`.
+    ///
+    /// If it is called the first time for `value`, it will possibly leak the value and return an
+    /// [`Interned<T>`] using the obtained static reference. Subsequent calls for the same `value`
+    /// will return [`Interned<T>`] using the same static reference.
+    pub fn intern<Q: Leak<T>>(&self, value: Q) -> Interned<T> {
+        let lock = self.0.get_or_init(Default::default);
+        {
+            let set = lock.read().unwrap_or_else(PoisonError::into_inner);
+            if let Some(value) = set.get(value.borrow()) {
+                return Interned(*value);
+            }
+        }
+        {
+            let mut set = lock.write().unwrap_or_else(PoisonError::into_inner);
+            if let Some(value) = set.get(value.borrow()) {
+                Interned(*value)
+            } else {
+                let leaked = value.leak();
+                set.insert(leaked);
+                Interned(leaked)
+            }
+        }
+    }
+}
+
+impl<T: ?Sized> Default for Interner<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A type that can provide static references to equal values.
+pub trait StaticRef {
+    /// Returns a static reference to a value equal to `self`, if possible.
+    /// This method is used by [`OptimizedInterner::intern`] to optimize the interning process.
+    ///
+    /// # Invariant
+    ///
+    /// The following invariant must be hold:
+    ///
+    /// `ptr_eq(a.static_ref(), b.static_ref()) == true` if and only if `a == b`
+    ///
+    /// where `ptr_eq` is defined as :
+    /// ```
+    /// fn ptr_eq<T>(x: Option<&'static T>, y: Option<&'static T>) -> bool {
+    ///     match (x, y) {
+    ///         (Some(x), Some(y)) => std::ptr::eq(x, y),
+    ///         _ => false,
+    ///     }
+    /// }
+    /// ```
```

Comment:
```suggestion
    /// The following invariant must be hold:
    ///
    /// `ptr_eq(a.static_ref(), b.static_ref()) == true` only if `a == b`
    ///
    /// where `ptr_eq` is defined as :
    /// ```
    /// fn ptr_eq<T>(x: Option<&'static T>, y: Option<&'static T>) -> bool {
    ///     match (x, y) {
    ///         (Some(x), Some(y)) => std::ptr::eq(x, y),
    ///         _ => false,
    ///     }
    /// }
    /// ```
```

I believe the "if and only if" choice of words is slightly incorrect here. If `static_ref` returns `None`, then `ptr_eq` well evaluate to `false`, but `a` can still be equal to `b`.

Reply:
You are right, but the invariants are a bit more complex:

Basically, if we return `None` we are always safe. In the cases we return `Some`, we not only need to make sure that the pointer is identical for all equal values but also that the pointer is different for unequal values. 
And we cannot mix `Some` and `None` for equal values.

I resolved this by splitting the invariant into two parts.